text/006-bft-based-ordering-service.md

```diff
@@ -0,0 +1,405 @@
+The `prepare` and `commit` phases are used to ensure that requests that commit are totally ordered across views. 
+
+![alt text](https://raw.githubusercontent.com/SmartBFT-Go/SmartBFT-Go.github.io/master/flow.png)
+
+
+In the `pre-prepare` phase, the leader batches requests to assemble a proposal, and it then assigns a sequence number to the proposal. Next the leader sends a `pre-prepare` message to all followers containing the proposal, its sequence number and the current view number `v`.
+
+A follower accepts the `pre-prepare` message if the proposal is valid, it is in view `v`, it is expecting this sequence number, and it has not accepted a different `pre-prepare` message for this view and sequence number.
+Once a follower accepts a `pre-prepare` message it enters the `prepare` phase by broadcasting a `prepare` message.
+The `prepare` message contains a digest of the accepted proposal, the sequence number, and `v`.
+
+Next each node waits for a quorum of `prepare` messages with the same digest, view `v`, and sequence number. 
+The quorum size should be big enough to ensure that in the intersection of any two quorums there is at least one correct node. So for example, if `n=3f+1` then the optimal quorum size is `2f+1`. 
+
+After the node is "prepared" it sends a `commit` message with the same content as the `prepare` message, as well as a signature on the proposal.
+Each node collects a quorum of `commit` messages and their attached signatures. 
+These signatures are later used to notarize the proposal, essentially creating a checkpoint. 
+Upon receiving a quorum of validated `commit` messages, the node delivers its proposal and the signatures to the application layer for storage and processing.
+
+The quorum size is calculated in the following manner: Let `n` be the total number of nodes in the algorithm and denote `f = ((n - 1) / 3)`. Then, the quorum is: `Ceil((N + F + 1) / 2)` where `Ceil` is the ceiling function. In layman's terms, the quorum is the minimal set of nodes such that every two such sets intersect in at least 1 honest node.  
+
+**View change**:
+The view change protocol is inspired by the _Synchronization Phase_ in this [BFT-SMaRt paper](http://www.di.fc.ul.pt/~bessani/publications/edcc12-modsmart.pdf) (illustrated in the attached figure taken from the BFT-SMaRt paper).
+The view change protocol is comprised of three messages: `ViewChange`, `ViewData`, and `NewView` (`Stop`, `StopData`, and `Sync` in BFT-SMaRt).
+
+![alt text](https://raw.githubusercontent.com/wiki/SmartBFT-Go/consensus/images/viewChange.png)
+
+The `ViewChange` message (the first message) is sent to all nodes by some node who suspects the leader is faulty. This message is very lightweight and includes only the next view sequence number. 
+Once collecting `2f+1` `ViewChange` messages, a non faulty node is convinced that enough nodes consider the leader as faulty, and it will send a `ViewData` message to the next view's leader.
```

Comment:
How do the non-faulty nodes agree on which node should be the next leader?

Reply:
In the classical BFT there is no "agreement" on the next leader, but it rotates to the next node in accordance to its numerical identifier. I added a clarification.

Reply:
Also, any BFT protocol that lets the nodes "elect" the leader in a non-randomized process would lead to an attack where an adversary that controls several byzantine nodes constantly makes its nodes be re-elected again as leaders. 
Unless you take some sort of ["black-listing" approach](https://patents.google.com/patent/US11343313B1/en), or a real [randomized approach](https://eprint.iacr.org/2020/025) (which is computationally expensive)

Reply:
> any BFT protocol that lets the nodes "elect" the leader in a non-randomized process would lead to an attack where an adversary that controls several byzantine nodes constantly makes its nodes be re-elected again as leaders.

Could you please share a literature reference or elaborate a little more yourself to support this claim?

How rotation according to nodes' numerical identifier is different from the non-randomized leader selection process?

Reply:
>  Could you please share a literature reference or elaborate a little more yourself to support this claim?
> 
> How rotation according to nodes' numerical identifier is different from the non-randomized leader selection process?

So, take for example what happens with Raft. And I suppose this is also why Andrew asked this question.
In Raft, the leader election doesn't use rotation like in BFT, but the candidates send out requests to other candidates to vote for them, and eventually they gather a majority of votes. 
A Byzantine candidate may always act faster than everyone else and send quickly requests for votes from everyone else. 

In a (truly) randomized process (where the randomness is verifiable and not just randomized in the honest setting) we can think of the the leader election process as a deterministic function where the inputs are random and come from the candidates.